// @aztec/bb.js/dest/node-cjs/types/point.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Point = void 0;
const index_js_1 = require("./index.js");
const buffer_reader_js_1 = require("../serialize/buffer_reader.js");
class Point {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
    static random() {
        // TODO: This is not a point on the curve!
        return new Point(index_js_1.Fr.random(), index_js_1.Fr.random());
    }
    static fromBuffer(buffer) {
        const reader = buffer_reader_js_1.BufferReader.asReader(buffer);
        return new this(index_js_1.Fr.fromBuffer(reader), index_js_1.Fr.fromBuffer(reader));
    }
    static fromString(address) {
        return Point.fromBuffer(Buffer.from(address.replace(/^0x/i, ''), 'hex'));
    }
    toBuffer() {
        return Buffer.concat([this.x.toBuffer(), this.y.toBuffer()]);
    }
    toString() {
        return '0x' + this.toBuffer().toString('hex');
    }
    equals(rhs) {
        return this.x.equals(rhs.x) && this.y.equals(rhs.y);
    }
}
exports.Point = Point;
Point.SIZE_IN_BYTES = 64;
Point.EMPTY = new Point(index_js_1.Fr.ZERO, index_js_1.Fr.ZERO);
