// @aztec/bb.js/dest/node-cjs/types/fixed_size_buffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Buffer128 = exports.Buffer64 = exports.Buffer32 = void 0;
const index_js_1 = require("../random/index.js");
const index_js_2 = require("../serialize/index.js");
class Buffer32 {
    constructor(buffer) {
        this.buffer = buffer;
    }
    static fromBuffer(buffer) {
        const reader = index_js_2.BufferReader.asReader(buffer);
        return new Buffer32(reader.readBytes(this.SIZE_IN_BYTES));
    }
    static random() {
        return new Buffer32((0, index_js_1.randomBytes)(this.SIZE_IN_BYTES));
    }
    toBuffer() {
        return this.buffer;
    }
}
exports.Buffer32 = Buffer32;
Buffer32.SIZE_IN_BYTES = 32;
class Buffer64 {
    constructor(buffer) {
        this.buffer = buffer;
    }
    static fromBuffer(buffer) {
        const reader = index_js_2.BufferReader.asReader(buffer);
        return new Buffer64(reader.readBytes(this.SIZE_IN_BYTES));
    }
    static random() {
        return new Buffer64((0, index_js_1.randomBytes)(this.SIZE_IN_BYTES));
    }
    toBuffer() {
        return this.buffer;
    }
}
exports.Buffer64 = Buffer64;
Buffer64.SIZE_IN_BYTES = 64;
class Buffer128 {
    constructor(buffer) {
        this.buffer = buffer;
    }
    static fromBuffer(buffer) {
        const reader = index_js_2.BufferReader.asReader(buffer);
        return new Buffer128(reader.readBytes(this.SIZE_IN_BYTES));
    }
    static random() {
        return new Buffer128((0, index_js_1.randomBytes)(this.SIZE_IN_BYTES));
    }
    toBuffer() {
        return this.buffer;
    }
}
exports.Buffer128 = Buffer128;
Buffer128.SIZE_IN_BYTES = 128;
