// @aztec/bb.js/dest/node-cjs/crs/node/ignition_files_crs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IgnitionFilesCrs = exports.GRUMPKIN_SRS_DEV_PATH = exports.SRS_DEV_PATH = void 0;
const fs_1 = require("fs");
const promises_1 = require("fs/promises");
const path_1 = require("path");
const url_1 = require("url");
function getCurrentDir() {
    if (typeof __dirname !== 'undefined') {
        return __dirname;
    }
    else {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        return (0, path_1.dirname)((0, url_1.fileURLToPath)(""));
    }
}
/**
 * The path to our SRS object, assuming that we are in e.g. barretenberg/ts/dest/node/crs/node folder.
 */
exports.SRS_DEV_PATH = getCurrentDir() + '/../../../../../cpp/srs_db/ignition/monomial';
exports.GRUMPKIN_SRS_DEV_PATH = getCurrentDir() + '/../../../../../cpp/srs_db/grumpkin/monomial';
/**
 * Downloader for CRS from a local file (for Node).
 */
class IgnitionFilesCrs {
    constructor(
    /**
     * The number of circuit gates.
     */
    numPoints, path = exports.SRS_DEV_PATH) {
        this.numPoints = numPoints;
        this.path = path;
    }
    pathExists() {
        return (0, fs_1.existsSync)(this.path);
    }
    /**
     * Read the data file.
     */
    async init() {
        // We need this.numPoints number of g1 points.
        // numPoints should be circuitSize + 1.
        const g1Start = 28;
        const g1End = g1Start + this.numPoints * 64;
        const data = await (0, promises_1.readFile)(this.path + '/transcript00.dat');
        this.data = data.subarray(g1Start, g1End);
        // TODO(https://github.com/AztecProtocol/barretenberg/issues/811): proper abstraction from Grumpkin which does not have g2
        if ((0, fs_1.existsSync)(this.path + '/g2.dat')) {
            this.g2Data = await (0, promises_1.readFile)(this.path + '/g2.dat');
        }
    }
    /**
     * G1 points data for prover key.
     * @returns The points data.
     */
    getG1Data() {
        return this.data;
    }
    /**
     * G2 points data for verification key.
     * @returns The points data.
     */
    getG2Data() {
        return this.g2Data;
    }
}
exports.IgnitionFilesCrs = IgnitionFilesCrs;
