// @aztec/bb.js/dest/node-cjs/barretenberg_wasm/barretenberg_wasm_thread/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BarretenbergWasmThread = void 0;
const index_js_1 = require("../helpers/index.js");
const index_js_2 = require("../barretenberg_wasm_base/index.js");
class BarretenbergWasmThread extends index_js_2.BarretenbergWasmBase {
    /**
     * Init as worker thread.
     */
    async initThread(module, memory) {
        this.logger = (0, index_js_1.threadLogger)() || this.logger;
        this.memory = memory;
        this.instance = await WebAssembly.instantiate(module, this.getImportObj(this.memory));
    }
    destroy() {
        (0, index_js_1.killSelf)();
    }
    getImportObj(memory) {
        const baseImports = super.getImportObj(memory);
        /* eslint-disable camelcase */
        return {
            ...baseImports,
            wasi: {
                'thread-spawn': () => {
                    this.logger('PANIC: threads cannot spawn threads!');
                    this.logger(new Error().stack);
                    (0, index_js_1.killSelf)();
                },
            },
            // These are functions implementations for imports we've defined are needed.
            // The native C++ build defines these in a module called "env". We must implement TypeScript versions here.
            env: {
                ...baseImports.env,
                env_hardware_concurrency: () => {
                    // We return 1, which should cause any algos using threading to just not create a thread.
                    return 1;
                },
            },
        };
        /* eslint-enable camelcase */
    }
}
exports.BarretenbergWasmThread = BarretenbergWasmThread;
