// @aztec/bb.js/dest/node-cjs/serialize/output_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringDeserializer = exports.BufferDeserializer = exports.VectorDeserializer = exports.NumberDeserializer = exports.BoolDeserializer = void 0;
const buffer_reader_js_1 = require("./buffer_reader.js");
function BoolDeserializer() {
    return {
        SIZE_IN_BYTES: 1,
        fromBuffer: (buf) => {
            const reader = buffer_reader_js_1.BufferReader.asReader(buf);
            return reader.readBoolean();
        },
    };
}
exports.BoolDeserializer = BoolDeserializer;
function NumberDeserializer() {
    return {
        SIZE_IN_BYTES: 4,
        fromBuffer: (buf) => {
            const reader = buffer_reader_js_1.BufferReader.asReader(buf);
            return reader.readNumber();
        },
    };
}
exports.NumberDeserializer = NumberDeserializer;
function VectorDeserializer(t) {
    return {
        fromBuffer: (buf) => {
            const reader = buffer_reader_js_1.BufferReader.asReader(buf);
            return reader.readVector(t);
        },
    };
}
exports.VectorDeserializer = VectorDeserializer;
function BufferDeserializer() {
    return {
        fromBuffer: (buf) => {
            const reader = buffer_reader_js_1.BufferReader.asReader(buf);
            return reader.readBuffer();
        },
    };
}
exports.BufferDeserializer = BufferDeserializer;
function StringDeserializer() {
    return {
        fromBuffer: (buf) => {
            const reader = buffer_reader_js_1.BufferReader.asReader(buf);
            return reader.readString();
        },
    };
}
exports.StringDeserializer = StringDeserializer;
